x-pack/plugins/monitoring/server/alerts/disk_usage_alert.ts

```diff
@@ -0,0 +1,403 @@
+        {
+          defaultMessage: `Disk usage alert is firing for {count} node(s) in cluster: {clusterName}. {shortActionText}`,
+          values: {
+            count: firingCount,
+            clusterName: cluster.clusterName,
+            shortActionText,
+          },
+        }
+      );
+      const internalFullMessage = i18n.translate(
+        'xpack.monitoring.alerts.diskUsage.firing.internalFullMessage',
+        {
+          defaultMessage: `Disk usage alert is firing for {count} node(s) in cluster: {clusterName}. {action}`,
+          values: {
+            count: firingCount,
+            clusterName: cluster.clusterName,
+            action,
+          },
+        }
+      );
+
+      instance.scheduleActions('default', {
+        internalShortMessage,
+        internalFullMessage: this.isCloud ? internalShortMessage : internalFullMessage,
+        state: AlertingDefaults.ALERT_STATE.firing,
+        nodes: firingNodes
+          .map((state) => `${state.nodeName}:${state.diskUsage.toFixed(2)}`)
+          .join(','),
+        count: firingCount,
+        clusterName: cluster.clusterName,
+        action,
+        actionPlain: shortActionText,
+      });
+    } else {
+      const resolvedNodes = (alertStates as AlertDiskUsageState[])
+        .filter((state) => !state.ui.isFiring)
+        .map((state) => `${state.nodeName}:${state.diskUsage.toFixed(2)}`);
+      const resolvedCount = resolvedNodes.length;
+
+      if (resolvedCount > 0) {
+        const internalMessage = i18n.translate(
+          'xpack.monitoring.alerts.diskUsage.resolved.internalMessage',
+          {
+            defaultMessage: `Disk usage alert is resolved for {count} node(s) in cluster: {clusterName}.`,
+            values: {
+              count: resolvedCount,
+              clusterName: cluster.clusterName,
+            },
+          }
+        );
+
+        instance.scheduleActions('default', {
+          internalShortMessage: internalMessage,
+          internalFullMessage: internalMessage,
+          state: AlertingDefaults.ALERT_STATE.resolved,
+          nodes: resolvedNodes.join(','),
+          count: resolvedCount,
+          clusterName: cluster.clusterName,
+        });
+      }
+    }
+  }
+
+  private executeDeltas(
+    services: AlertServices,
+    cluster: AlertCluster,
+    newAlertStates: AlertDiskUsageState[],
+    oldAlertStates: AlertDiskUsageState[]
+  ) {
+    const deltaFiringStates = [];
+    const deltaResolvedStates = [];
+    const deltaInstanceIdPrefix: string = `.monitoring:${this.type}:${
+      cluster.clusterUuid
+    }:${Date.now()}:`;
+
+    for (const newAlertState of newAlertStates) {
+      const relatedOldState = oldAlertStates.find(
+        (oldState) =>
+          oldState.nodeId === newAlertState.nodeId &&
+          oldState.ui.isFiring !== newAlertState.ui.isFiring &&
+          oldState.ui.resolvedMS !== newAlertState.ui.resolvedMS
+      );
+      if (!relatedOldState) {
```

Comment:
I'm not sure I understand this logic here.

If the the old state is the same node as the new state, but the `isFiring` flipped and `resolvedMS` is different, that means we need to fire a resolution? I'm not sure we even need to look at `resolvedMS`. If the `isFiring` went from `true` to `false`, then we need to fire resolved actions.

Or am I just not reading this properly?